### ISSUE TITLE
Generate model frame for multiple observations

### ISSUE THREAD

**Issue by pmelchior** [pmelchior/scarlet#156]
We currently have a union/intersection computation in `resampling.match_patches`, which is called by `LowResObservation`. This is too implicit.

In #120, I suggested a class method:
```
frame = Frame.from_observations(observations, coverage='intersection')
```
where `coverage in ['intersection', 'union']`. This method *may* also decide what wcs and PSF is should have, e.g. by simply picking the finest-resolution wcs and the sharpest PSF. It can also automatically match all the frames to the model frame.

I prefer the classmethod here because there is no other functionality beside the initial choice of footprints that cannot be done by the default `Frame` implementation. That's ideal for a class method. It would also align with another useful class method
```
frame = Frame.from_survey("HSC", tile_id=104928)
```
for scanning a full survey one frame at a time.

**Comment by pmelchior** [comment]
The difficulty with this approach is that the model frame needs to be wider on the sky than the observations because of the blurring from the diff kernel. Irrespective of union/intersection, we need to "pad" the model frame to allow for the widest diff kernel of any observation.

**Comment by herjy** [comment]
So it is essentially the same issue as #120. But this is almost a semantic choice do we decide of the extent of the fitting when declaring the frame or when matching observations to the frame. To me it made slightly more sense to have this in the matching because matching is, in my head, where we should becide this. 
Also, if we were to have multiple `LowResObservations`, it would allow us to decide whether we want each intersection or overlap independently for each instance. I don't see an obvious reason to absolutely have that possibility, but I don't see anything against it either.

**Comment by pmelchior** [comment]
It can't be in `match` because that has access only the the model frame and the properties of this single observation, while the process I describe needs to know about multiple observations.

It is a bit circular: in order to define the model frame, we need to know the observed frames, and then we match the observed frames to the model frame. However, this split means that I can deal with random pointings and with contiguous survey areas.

Also, by the same argument about the diff kernel width, we also need to adjust the treatment for a single `Observation`. We currently only pad, but the pad is empty, while the true sky can have sources out there, whose flux will "leak" into the observed frame.

**Comment by fred3m** [comment]
I don't think this will affect anything that I'm doing, so I don't really have a dog in this fight or feel strongly about it, this is just a suggestion:

This depends on how you want to define your model frame, right? If you're creating a model frame `from_observations` then you're making a decision for all users about how they will want to create the model PSF. That might be ok, but it might be that they want to use an analytic PSF and not base that on the observed PSFs, which should in principle be ok as long as the analytic PSF is narrower than all of the observed PSFs.

So I agree with @pmelchior that we should be able to match multiple observations to a frame simultaneously but I also think that the model frame should have the ability to be generated outside of observation frames and I agree with @herjy that `match` is the more natural place for this calculation.

So just to muddy the water I have a proposal: what if instead of calling `match` from each `observation` we call it from the model `frame`, and pass the observation(s) that we are going to match? For example initialization in the multi-resolution tutorial would become:

```
frame = scarlet.Frame(shape, wcs=wcs_hst, psfs=psf_hst, channels=channels)

# define two observation packages and match to frame
obs_hst = scarlet.Observation(data_hst, wcs=wcs_hst,psfs=psf_hst,
                              channels=channels_hst, weights=weights_hst)
obs_hsc = scarlet.LowResObservation(data_hsc,  wcs=wcs_hsc, psfs=psf_hsc,
                                    channels=channels_hsc, weights=weights_hsc)
frame.match([obs_hst, obs_hsc])
```

This allows us to specify our model and observations independently (if we so desire) but then will generate the difference kernel in each observation. In the case where one of the observations has a PSF that matches the frame PSF, the existing code should prevent the creation of the difference kernel.

**Comment by pmelchior** [comment]
I agree: central to this code is the flexibility to decide what exactly the model frame is (which is why I think of two classmethods). This issue is specifically about the shape of the model frame. @fred3m Your point about the PSF of that frame is a good catch because the union/intersection argument doesn't address it at all.

@herjy and I had a brief conversation. The first aspect is conceptual. I see `match` as fully determined by the model frame and the observed frame. Given a model frame, it explains how to get to the observed data. This mapping requires knowledge of the PSF, resolution, filters, etc., it depends on things that are stored in Frame, so it could be a method of Frame, as @fred3m suggested. But that would make it ambiguous: `frame.match(other_frame)` does not state whose properties are being set here. Currently, that is clear.

The second is a practical approach to implement `from_observations`:

1. From all observation WCSs: adopt highest resolution (unless given by argument `wcs`)
2. From all observation PSFs: adopt sharpest (unless given by argument `psfs`). This needs a size measure like `fwhm` in `PSF`.
3. Check if model PSF is well sampled in model pixels, issue warning if not. If OK, render sharpest PSF in pixels of the model frame. Check if identical to one of the observations, if so use reference to it as PSF
4. From the observed WCSs compute union/intersection to set model shape.
5. Extend model shape in spatial direction by shape of the widest PSF image (in model pixels) or by width of the widest PSF (again in model pixels).

This approach will work for single observations like LSST or for multiple observations.

Somewhere in there (after step 3) we should do the astrometric correction we talked about, but I'm not yet convinced how.

**Comment by herjy** [comment]
I did not have a STRONG feeling about it either, the discussion allowed me to better understand the mindset behind frame and observations. I like that solution, because it still makes frame the place where we build the model frame entirely. I guess frame.match would call the match of the observations, but then all the mechanics that does the coordinate matching would be methods of frame

**Comment by pmelchior** [comment]
@herjy you like which solution exactly? 

My point is making `match` a method of Frame is redundant (because Observation has a Frame) and ambiguous (see above).

**Comment by herjy** [comment]
@fred3m 's I don't see the issue. Observation inherits from frame so the redundancy does not bother me and we don't need to run `obs_frame.match()` just the same way as we would not run `obs_frame.from_observation(potato)`which is just as redundant. You are proposing the same with different names in the end.

**Comment by herjy** [comment]
Coming back to 3, the frame psf does not need to be well sampled in the model frame if I'm not mistaken. All the others do though. 
I like being able to have an automated setup for the frame, but at the same time It makes me feel like inviting the user to skip a crucial concept to understand which is that of the model frame.

**Comment by pmelchior** [comment]
> Observation inherits from frame so the redundancy does not bother me and we don't need to run `obs_frame.match()` just the same way as we would not run `obs_frame.from_observation(potato)`which is just as redundant. You are proposing the same with different names in the end.

Few things: `Observation` does not inherit from `Frame`, it has a `Frame`. More importantly, if the matching code is in `Frame`, the code does not specify in which direction the matching happens. As long as it's a method of `Observation` there's no ambiguity. Also, I think you still have trouble with `@classmethod`. It will generate the Frame that suitably comprises the observations.

**Comment by pmelchior** [comment]
> Coming back to 3, the frame psf does not need to be well sampled in the model frame if I'm not mistaken. All the others do though.

The model PSF needs to be well sampled, otherwise none of the math goes through.

> I like being able to have an automated setup for the frame, but at the same time It makes me feel like inviting the user to skip a crucial concept to understand which is that of the model frame.

I feel the same conflict, but I lean to codifying best practices. User should be able to deviate if needed.

**Comment by herjy** [comment]
> The model PSF needs to be well sampled, otherwise none of the math goes through.
No, because we never interpolate the high resolution psf or deconvolve by it. It is only the diff kernel which needs to be well sampled.

**Comment by pmelchior** [comment]
It's not about the practicalities. The equations we use for the sampling+convolution are only valid for well-sampled PSFs. Otherwise the convolution and the sampling don't commute.

**Comment by pmelchior** [comment]
Can we just try it with the `@classmethod` approach I laid out above. This will work for the immediate needs. This method may or may not do relative astrometric optimization, but it should be able to do the some shape padding based on the size of the widest PSF. I'm thinking that without creating a measured PSF width, we can adopt something like half of the PSF image (in units of the model pixels).

**Comment by fred3m** [comment]
I'm not sure I understand how the `from_observations` constructor is intended to work and cover all of the things that we discussed above, did you two have a discussion offline? How will the model PSF be initialized, the orientation (WCS) be determined, and the shape be determined? It seems like this method will do everything that `match` does, so will `match` be removed from `Observation`, or will it remain in observation and be called by `from_observations`?

**Comment by pmelchior** [comment]
@fred3m Good questions. Before getting into details, I don't want to replace `Observation.match`. Instead I want to make it simple. Once the frame of the model is clear, `match` should be fully specified, i.e. not change aspects of the model frame as a side effect. That's why I proposed to have `@classmethod` implementation for common scenarios, e.g. combining a few pointed observations or scanning a survey area.

For the multi-observation case:

* PSF should be a Gaussian with a minimum width so to be well-sampled in the model frame (this is why it's step 2 in the list above)
* WCS (the transformation, not the shape) is probably identical to the highest res observation, so that one observation does not have to be resampled.
* Shape is determined based on the coverage of the data sets on the sky as a bounding box for the intersection or the union.

The last item started it all because it's currently done as part of `LowResObservation.match`, which is the wrong place because it's creating side effects that propagate "backwards" to the model. I think we should be declarative: 1) This is the model, 2) this is how you match it to observations.

**Comment by herjy** [comment]
One of the problems that arises here comes from the dichotomia between Observations and LowResObservations. The user still has to know in advance what Observation class to use and this will depend on whether the wcs matches. My current thought is that once the target wcs is identifed, we make all observations that have a different wcs a LowResObservation :/

**Comment by pmelchior** [comment]
Good point. So, we could do this:
```
obs1 = Observation(...)
obs2 = Observation(...)
model_frame = Frame.from_observations((obs1, obs2), coverage='union')
```
after which, say, `obs2` is turned into `LowResObservation` (maybe by `from_observations`), but it would not be obvious to the user.

**Comment by pmelchior** [comment]
I believe this has been closed by 9a19f6c. @herjy please confirm and close this issue if true

**Comment by herjy** [comment]
It most definitely has!
